TestRegistryEvents is flaky

```
2015/07/08 11:32:42 docker container: collecting stats for ping
2015/07/08 11:32:42 docker container: error reading event io: read/write on closed pipe
2015/07/08 11:32:42 docker container: stopped collecting stats for ping
--- FAIL: TestRegistryEvents-4 (0.00s)
	registry_test.go:196: 
		--- want
		+++ have
		@@ -1,53 +1,3 @@
		-([]docker.Container) (len=2 cap=2) {
		- (*docker_test.mockContainer)(0xc2080ee018)({
		-  c: (*docker.Container)(0x14b4440)({
		-   ID: (string) (len=4) "ping",
		-   Created: (time.Time) 0001-01-01 00:00:00 +0000 UTC,
		-   Path: (string) "",
		-   Args: ([]string) <nil>,
		-   Config: (*docker.Config)(<nil>),
		-   State: (docker.State) Up 2562047h47m16.854775807s,
		-   Image: (string) (len=3) "baz",
		-   Node: (*docker.SwarmNode)(<nil>),
		-   NetworkSettings: (*docker.NetworkSettings)(<nil>),
		-   SysInitPath: (string) "",
		-   ResolvConfPath: (string) "",
		-   HostnamePath: (string) "",
		-   HostsPath: (string) "",
		-   Name: (string) (len=4) "pong",
		-   Driver: (string) "",
		-   Volumes: (map[string]string) <nil>,
		-   VolumesRW: (map[string]bool) <nil>,
		-   HostConfig: (*docker.HostConfig)(<nil>),
		-   ExecIDs: ([]string) <nil>,
		-   RestartCount: (int) 0,
		-   AppArmorProfile: (string) ""
		-  })
		- }),
		- (*docker_test.mockContainer)(0xc2080ee020)({
		-  c: (*docker.Container)(0x14b45c0)({
		-   ID: (string) (len=4) "wiff",
		-   Created: (time.Time) 0001-01-01 00:00:00 +0000 UTC,
		-   Path: (string) "",
		-   Args: ([]string) <nil>,
		-   Config: (*docker.Config)(<nil>),
		-   State: (docker.State) Up 2562047h47m16.854775807s,
		-   Image: (string) (len=3) "baz",
		-   Node: (*docker.SwarmNode)(<nil>),
		-   NetworkSettings: (*docker.NetworkSettings)(<nil>),
		-   SysInitPath: (string) "",
		-   ResolvConfPath: (string) "",
		-   HostnamePath: (string) "",
		-   HostsPath: (string) "",
		-   Name: (string) (len=4) "waff",
		-   Driver: (string) "",
		-   Volumes: (map[string]string) <nil>,
		-   VolumesRW: (map[string]bool) <nil>,
		-   HostConfig: (*docker.HostConfig)(<nil>),
		-   ExecIDs: ([]string) <nil>,
		-   RestartCount: (int) 0,
		-   AppArmorProfile: (string) ""
		-  })
		- })
		+([]docker.Container) {
		 }
		 
FAIL
```

```
2015/07/08 12:10:08 docker container: collecting stats for ping
2015/07/08 12:10:08 docker container: error reading event io: read/write on closed pipe
2015/07/08 12:10:08 docker container: stopped collecting stats for ping
panic: test timed out after 10s

goroutine 19 [running]:
testing.func·008()
	/usr/local/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive]:
testing.RunTests(0x9de188, 0x14b3e40, 0x5, 0x5, 0xc208078401)
	/usr/local/go/src/testing/testing.go:556 +0xdc3
testing.(*M).Run(0xc208030550, 0x14bf660)
	/usr/local/go/src/testing/testing.go:485 +0xe1
main.main()
	github.com/weaveworks/scope/probe/docker/_test/_testmain.go:114 +0x382

goroutine 17 [chan send]:
github.com/weaveworks/scope/probe/docker_test.(*mockDockerClient).send(0xc2080e8060, 0xc2080ce2c0)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:92 +0x104
github.com/weaveworks/scope/probe/docker_test.func·012()
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:190 +0x1d3
github.com/weaveworks/scope/probe/docker_test.setupStubs(0xc2080e8060, 0xc2080f2750)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:130 +0x1dc
github.com/weaveworks/scope/probe/docker_test.TestRegistryEvents(0xc208080120)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:238 +0x102
testing.tRunner(0xc208080120, 0x14b3e70)
	/usr/local/go/src/testing/testing.go:447 +0x134
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xd4e

goroutine 18 [semacquire]:
sync.(*Mutex).Lock(0xc2080e8060)
	/usr/local/go/src/sync/mutex.go:66 +0xfd
github.com/weaveworks/scope/probe/docker_test.(*mockDockerClient).ListImages(0xc2080e8060, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:65 +0x6e
github.com/weaveworks/scope/probe/docker.(*registry).updateImages(0xc2080f8000, 0x0, 0x0)
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:206 +0xf5
github.com/weaveworks/scope/probe/docker.(*registry).listenForEvents(0xc2080f8000, 0x0)
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:128 +0x4a9
github.com/weaveworks/scope/probe/docker.(*registry).loop(0xc2080f8000)
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:90 +0x6e
created by github.com/weaveworks/scope/probe/docker.NewRegistry
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:74 +0x38b
FAIL	github.com/weaveworks/scope/probe/docker	10.017s
ok  	github.com/weaveworks/scope/probe/endpoint	0.022s	coverage: 53.7% of statements
ok  	github.com/weaveworks/scope/probe/host	0.017s	coverage: 78.3% of statements
ok  	github.com/weaveworks/scope/probe/overlay	0.017s	coverage: 67.7% of statements
ok  	github.com/weaveworks/scope/probe/process	0.016s	coverage: 89.4% of statements
ok  	github.com/weaveworks/scope/render	0.030s	coverage: 91.7% of statements
ok  	github.com/weaveworks/scope/report	0.145s	coverage: 62.4% of statements
```

```
2015/07/08 12:54:57 docker container: collecting stats for ping
2015/07/08 12:54:57 docker container: error reading event io: read/write on closed pipe
2015/07/08 12:54:57 docker container: stopped collecting stats for ping
==================
WARNING: DATA RACE
Read by goroutine 11:
  runtime.mapaccess1_faststr()
      /usr/local/go/src/runtime/hashmap_fast.go:179 +0x0
  github.com/weaveworks/scope/probe/docker_test.(*mockDockerClient).InspectContainer()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:61 +0x92
  github.com/weaveworks/scope/probe/docker.(*registry).addContainer()
      github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:246 +0xc7
  github.com/weaveworks/scope/probe/docker.(*registry).updateContainers()
      github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:194 +0x276
  github.com/weaveworks/scope/probe/docker.(*registry).listenForEvents()
      github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:121 +0x321
  github.com/weaveworks/scope/probe/docker.(*registry).loop()
      github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:90 +0x6d

Previous write by goroutine 10:
  runtime.mapassign1()
      /usr/local/go/src/runtime/hashmap.go:376 +0x0
  github.com/weaveworks/scope/probe/docker_test.func·014()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:195 +0x1bd
  github.com/weaveworks/scope/probe/docker_test.setupStubs()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:130 +0x1db
  github.com/weaveworks/scope/probe/docker_test.TestRegistryEvents()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:233 +0x101
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133

Goroutine 11 (running) created at:
  github.com/weaveworks/scope/probe/docker.NewRegistry()
      github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:74 +0x38b
  github.com/weaveworks/scope/probe/docker_test.func·014()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:182 +0x5a
  github.com/weaveworks/scope/probe/docker_test.setupStubs()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:130 +0x1db
  github.com/weaveworks/scope/probe/docker_test.TestRegistryEvents()
      /home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:233 +0x101
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:447 +0x133

Goroutine 10 (running) created at:
  testing.RunTests()
      /usr/local/go/src/testing/testing.go:555 +0xd4e
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:485 +0xe0
  main.main()
      github.com/weaveworks/scope/probe/docker/_test/_testmain.go:114 +0x381
==================
panic: test timed out after 10s

goroutine 8 [running]:
testing.func·008()
	/usr/local/go/src/testing/testing.go:681 +0x168
created by time.goFunc
	/usr/local/go/src/time/sleep.go:129 +0x59

goroutine 1 [chan receive]:
testing.RunTests(0x9de808, 0x14b4e40, 0x5, 0x5, 0xc208078401)
	/usr/local/go/src/testing/testing.go:556 +0xdc3
testing.(*M).Run(0xc208030550, 0x14c0660)
	/usr/local/go/src/testing/testing.go:485 +0xe1
main.main()
	github.com/weaveworks/scope/probe/docker/_test/_testmain.go:114 +0x382

goroutine 33 [chan send]:
github.com/weaveworks/scope/probe/docker_test.(*mockDockerClient).send(0xc2080e0060, 0xc2080c4100)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:92 +0x104
github.com/weaveworks/scope/probe/docker_test.func·014()
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:197 +0x282
github.com/weaveworks/scope/probe/docker_test.setupStubs(0xc2080e0060, 0xc2080ec750)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:130 +0x1dc
github.com/weaveworks/scope/probe/docker_test.TestRegistryEvents(0xc2080e2000)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:233 +0x102
testing.tRunner(0xc2080e2000, 0x14b4e70)
	/usr/local/go/src/testing/testing.go:447 +0x134
created by testing.RunTests
	/usr/local/go/src/testing/testing.go:555 +0xd4e

goroutine 34 [semacquire]:
sync.(*Mutex).Lock(0xc2080e0060)
	/usr/local/go/src/sync/mutex.go:66 +0xfd
github.com/weaveworks/scope/probe/docker_test.(*mockDockerClient).ListImages(0xc2080e0060, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/home/ubuntu/src/github.com/weaveworks/scope/probe/docker/registry_test.go:65 +0x6e
github.com/weaveworks/scope/probe/docker.(*registry).updateImages(0xc2080f0000, 0x0, 0x0)
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:206 +0xf5
github.com/weaveworks/scope/probe/docker.(*registry).listenForEvents(0xc2080f0000, 0x0)
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:128 +0x4a9
github.com/weaveworks/scope/probe/docker.(*registry).loop(0xc2080f0000)
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:90 +0x6e
created by github.com/weaveworks/scope/probe/docker.NewRegistry
	github.com/weaveworks/scope/probe/docker/_test/_obj_test/registry.go:74 +0x38b
FAIL	github.com/weaveworks/scope/probe/docker	10.015s
```

--- a/probe/docker/container_test.go
+++ b/probe/docker/container_test.go
@@ -61,8 +61,7 @@ func TestContainer(t *testing.T) {
 	}
 
 	// Now see if we go them
-	test.Poll(t, 10*time.Millisecond, func() bool {
-		nmd := c.GetNodeMetadata()
-		return nmd[docker.MemoryUsage] == "12345"
-	}, "Failed to get stats")
+	test.Poll(t, 10*time.Millisecond, "12345", func() interface{} {
+		return c.GetNodeMetadata()[docker.MemoryUsage]
+	})
 }

--- a/probe/docker/registry_test.go
+++ b/probe/docker/registry_test.go
@@ -46,7 +46,7 @@ func (c *mockContainer) GetNodeMetadata() report.NodeMetadata {
 }
 
 type mockDockerClient struct {
-	sync.Mutex
+	sync.RWMutex
 	apiContainers []client.APIContainers
 	containers    map[string]*client.Container
 	apiImages     []client.APIImages
@@ -54,16 +54,20 @@ type mockDockerClient struct {
 }
 
 func (m *mockDockerClient) ListContainers(client.ListContainersOptions) ([]client.APIContainers, error) {
+	m.RLock()
+	defer m.RUnlock()
 	return m.apiContainers, nil
 }
 
 func (m *mockDockerClient) InspectContainer(id string) (*client.Container, error) {
+	m.RLock()
+	defer m.RUnlock()
 	return m.containers[id], nil
 }
 
 func (m *mockDockerClient) ListImages(client.ListImagesOptions) ([]client.APIImages, error) {
-	m.Lock()
-	defer m.Unlock()
+	m.RLock()
+	defer m.RUnlock()
 	return m.apiImages, nil
 }
 
@@ -86,8 +90,8 @@ func (m *mockDockerClient) RemoveEventListener(events chan *client.APIEvents) er
 }
 
 func (m *mockDockerClient) send(event *client.APIEvents) {
-	m.Lock()
-	defer m.Unlock()
+	m.RLock()
+	defer m.RUnlock()
 	for _, c := range m.events {
 		c <- event
 	}
@@ -160,11 +164,12 @@ func TestRegistry(t *testing.T) {
 		defer registry.Stop()
 		runtime.Gosched()
 
-		test.Poll(t, 10*time.Millisecond, func() bool {
-			have := allContainers(registry)
+		{
 			want := []docker.Container{&mockContainer{container1}}
-			return reflect.DeepEqual(want, have)
-		}, "Didn't get containers!")
+			test.Poll(t, 10*time.Millisecond, want, func() interface{} {
+				return allContainers(registry)
+			})
+		}
 
 		{
 			have := allImages(registry)
@@ -183,6 +188,12 @@ func TestRegistryEvents(t *testing.T) {
 		defer registry.Stop()
 		runtime.Gosched()
 
+		check := func(want []docker.Container) {
+			test.Poll(t, 10*time.Millisecond, want, func() interface{} {
+				return allContainers(registry)
+			})
+		}
+
 		{
 			mdc.Lock()
 			mdc.containers["wiff"] = container2
@@ -190,11 +201,8 @@ func TestRegistryEvents(t *testing.T) {
 			mdc.send(&client.APIEvents{Status: docker.StartEvent, ID: "wiff"})
 			runtime.Gosched()
 
-			have := allContainers(registry)
 			want := []docker.Container{&mockContainer{container1}, &mockContainer{container2}}
-			if !reflect.DeepEqual(want, have) {
-				t.Errorf("%s", test.Diff(want, have))
-			}
+			check(want)
 		}
 
 		{
@@ -204,11 +212,8 @@ func TestRegistryEvents(t *testing.T) {
 			mdc.send(&client.APIEvents{Status: docker.DieEvent, ID: "wiff"})
 			runtime.Gosched()
 
-			have := allContainers(registry)
 			want := []docker.Container{&mockContainer{container1}}
-			if !reflect.DeepEqual(want, have) {
-				t.Errorf("%s", test.Diff(want, have))
-			}
+			check(want)
 		}
 
 		{
@@ -218,22 +223,16 @@ func TestRegistryEvents(t *testing.T) {
 			mdc.send(&client.APIEvents{Status: docker.DieEvent, ID: "ping"})
 			runtime.Gosched()
 
-			have := allContainers(registry)
 			want := []docker.Container{}
-			if !reflect.DeepEqual(want, have) {
-				t.Errorf("%s", test.Diff(want, have))
-			}
+			check(want)
 		}
 
 		{
 			mdc.send(&client.APIEvents{Status: docker.DieEvent, ID: "doesntexist"})
 			runtime.Gosched()
 
-			have := allContainers(registry)
 			want := []docker.Container{}
-			if !reflect.DeepEqual(want, have) {
-				t.Errorf("%s", test.Diff(want, have))
-			}
+			check(want)
 		}
 	})
 }

--- a/test/poll.go
+++ b/test/poll.go
@@ -1,20 +1,25 @@
 package test
 
 import (
+	"reflect"
 	"testing"
 	"time"
 )
 
 // Poll repeatedly evaluates condition until we either timeout, or it suceeds.
-func Poll(t *testing.T, d time.Duration, condition func() bool, msg string) {
+func Poll(t *testing.T, d time.Duration, want interface{}, have func() interface{}) {
 	deadline := time.Now().Add(d)
 	for {
 		if time.Now().After(deadline) {
-			t.Fatal(msg)
+			break
 		}
-		if condition() {
+		if reflect.DeepEqual(want, have()) {
 			return
 		}
 		time.Sleep(d / 10)
+	}
+	h := have()
+	if reflect.DeepEqual(want, h) {
+		t.Fatal(Diff(want, h))
 	}
 }

--- a/xfer/collector_internal_test.go
+++ b/xfer/collector_internal_test.go
@@ -53,9 +53,9 @@ func TestCollector(t *testing.T) {
 	}
 
 	reports <- r
-	test.Poll(t, 100*time.Millisecond, func() bool {
-		return len(concreteCollector.peek().Address.NodeMetadatas) == 1
-	}, "missed the report")
+	test.Poll(t, 100*time.Millisecond, 1, func() interface{} {
+		return len(concreteCollector.peek().Address.NodeMetadatas)
+	})
 
 	go func() { publish <- time.Now() }()
 	collected := <-collector.Reports()

